### DISCUSSION TITLE
AA x AAPI communications for AANTS

### DISCUSSION THREAD

**Discussion by MinhxNguyen7** [icssc/AntAlmanac#1071] [Ideas]
We need to have a way for AAPI to ping us to check for changes in section enrollment. The previous understanding is that AAPI will send a list of courses that have been updated. We'd likely need an organization secret so that random people can't spam us.

@andrew-wang0, @IsaacNguyen will be principally developing AANTS on our side, and it'd be good if we can start talking about how this will work. I know that your team has other things on their plate, but this has been _the most_ requested feature for a long time, so it would be great if we can get the ball rolling to safely finish before next enrollment.

Related issues: #962, #971

**Comment by andrew-wang0** [comment]
Sounds good, currently AAPI is able to do a full scrape of websoc every ~5-10 minutes. (We're working on speeding this up, but for now this is the best we have for how often we can update enrollment statuses).

@IsaacNguyen, let me know if you have any ideas so far of how this could be implemented, but one possibility is setting up an endpoint that serves every section with a change in enrollment status that AA could poll.

**Reply by MinhxNguyen7** [reply]
We were thinking that AAPI would call an AA endpoint, instead of AA polling, since that'd be more wasteful.

With polling, what would the request/response look like?

**Reply by andrew-wang0** [reply]
Yes, I was thinking that would work too where AAPI could POST the data to AA so that AA wouldn't have to continuously poll.

However, there are a few reasons why I recommended it to be implemented the other way around:

- AAPI is already set up to serve GET requests (and AA to GET data from AAPI). It would be good to maintain this one-way connection rather than mixing them
- As an open source API, it would be beneficial to have the resource not tied specifically to AA and available to the public (or at least available through an API key)
- Polling eliminates the need for securing an endpoint on AA's side. Although a secret key could be used, the possibility of someone sending malicious notifications is a pretty big risk

Here is a rough idea of how a response from AAPI's side could look like:
```http 
GET https://anteaterapi.com/v2/rest/enrollmentChanges?sectionCodes=...&from=...&to=...
```
```json
{
  "ok": true,
  "data": {
    "courses": [
      {
        "deptCode": "BIO SCI",
        "courseTitle": "DNA TO ORGANISMS",
        "courseNumber": "93",
        "sections": [
          {
            "sectionCode": "12345",
            "maxCapacity": "75",
            "restrictionCodes": {
              "from": ["X", "Y"],
              "to": ["X"]
            },
            "status": {
              "from": "Waitl",
              "to": "OPEN"
            },
            "numCurrentlyEnrolled": {
              "totalEnrolled": {
                "from": "80",
                "to": "72"
              },
              "sectionEnrolled": {
                "from": "",
                "to": ""
              }
            },
            "numRequested": {
              "from": "85",
              "to": "78"
            },
            "numOnWaitlist": {
              "from": "3",
              "to": "0"
            },
            "numWaitlistCap": "10",
            "updatedAt": "2024-12-07T12:00:00Z"
          }
        ]
      }
    ]
  }
}
```

Every section with a change to its enrollment status (and every newly added section) since the last scrape would be included here. AA would have to poll until `updatedAt` changes, at which point it could parse the new data and send notifications to subscribed users. 

Let me know what you guys think.

**Reply by MinhxNguyen7** [reply]
> As an open source API, it would be beneficial to have the resource not tied specifically to AA and available to the public (or at least available through an API key)

I guess that's fair enough. I was thinking that, through a similar mechanism to API keys, we could add a URL to a list of URLs to be pinged when courses change, but that likely takes more combined effort and can induce a lot of load if there are a lot of consumers.

> Polling eliminates the need for securing an endpoint on AA's side. Although a secret key could be used, the possibility of someone sending malicious notifications is a pretty big risk.

The ping would just be for a list of courses that changed. We'd still query back the intersection between courses updated and subscribed. 

I think, then, going with polling would be the move.

**Reply by MinhxNguyen7** [reply]
That being said, I think it would be easier if we could just check the enrollment status of every section that people are subscribed to, then we compare the timestamps to see if we want to send a notification on our end.

**Reply by andrew-wang0** [reply]
Yep, we could make that a query parameter on the endpoint to filter for just the courses/sections that users are subscribed to

**Reply by MinhxNguyen7** [reply]
Example qeury: https://anteaterapi.com/v2/rest/enrollmentChanges?term=2024-WINTER&sections=123456,234567

**Comment by MinhxNguyen7** [comment]
@andrew-wang0 is there an issue up on the API side, and what would the timeline be for this?

**Reply by andrew-wang0** [reply]
Been discussing with @sanskarm7 about implementation details on our end. From the looks of it, we're thinking about a soft deadline of end of Winter quarter. 

Will put an issue up ASAP.

**Comment by IsaacNguyen** [comment]
Hi @andrew-wang0! Adding onto this, is there a possibility you guys could add functionality to detect when a class's major-only restriction is dropped to this endpoint? I'm unsure if something like this exists already in AAPI, but it would be a great feature if our users could also be notified when a class's restrictions are dropped. Thank you!

**Reply by andrew-wang0** [reply]
Hey @IsaacNguyen, unfortunately I don't believe that there is a universal way to scrape restriction drop timelines. Although it is listed on websoc, the format varies for each department (and sometimes even links to other websites or spreadsheets in order to get the info). 

For now, I don't think we're planning on implementing this.

**Reply by IsaacNguyen** [reply]
Okay that makes sense. Thank you!

**Reply by MinhxNguyen7** [reply]
The restriction *codes* are standardized, not the timelines. That's the same as class statuses.

@IsaacNguyen, there are more restrictions than just major restrictions. Click on the letter in the restriction column to see the definitions

**Reply by IsaacNguyen** [reply]
![IMG_1482](https://github.com/user-attachments/assets/9097e156-6dcb-45bc-9d70-31502016382a)
Besides major restrictions, which other restrictions have deadlines of when they’re dropped? The only ones here that make sense are the restrictions regarding a student’s class (freshman, sophomore, etc.) but I’m honestly unsure if those are dropped. Sorry I’m just not very familiar with how non-major restrictions work.

**Reply by MinhxNguyen7** [reply]
No problem. We don't know what the policy is, and they all work the sme way, so we can just implement all of them

**Comment by IsaacNguyen** [comment]
Hi @andrew-wang0! I was wondering what would be outputted from your end if a class has _not_ changed since the last scrape. Would it just show up with whatever data was updated since the last scrape? Also, during the period of time before any classes have any enrollment changes (like before any enrolllment windows have opened), what kind of output would I expect? Would it just be empty? Thank you!

**Reply by sanskarm7** [reply]
Hey Isaac, Jordan and I are currently working out the details as to how the variety of possible events will be addressed, and I think it'll be more solidified after we are able to meet with the AntAlmanac team as well. 

However, I do have an initial idea for how we can do this. Through consistently updating an additional table with only the recently updated information, while still having a normal "daily" update to the enrollment history table, we allow for the notification system to only have to deal with the newest information.

This would mean that a call that you make to get the newest data will only return the changed data. If nothing was changed since the last scrape, nothing will be returned. As such, if enrollment windows haven't been opened, there will be no changes, so it will be an empty output.

**Comment by laggycomputer** [comment]
Following up as I've joined this issue. I am now responsible for determining the output schema. I'm going to write a relatively naive endpoint. The model in https://github.com/icssc/AntAlmanac/discussions/1071#discussioncomment-11496536 is largely correct pending small changes, except that each section object will contain a `to` and optional `from`, each an object with fields describing the enrollment numbers.
I plan to only guarantee the following:
1. If both are present, `to` is the most recent scrape and `from` is the most recent scrape \*before* the time you provide. If no `from` snapshot meets that definition, it will not be present.
2. If both are present, they are distinct snapshots at different times and `from` was taken before `to`. If no `from` meets both conditions, it will not be present.

It is up to you to compare the snapshots to see if anything you care about has changed.

**Reply by MinhxNguyen7** [reply]
For us, I believe that it's fine even if you just tell us the change from the scrape before the most recent one to the most recent one, along with when the last scrape was (per section).

CC: @IsaacNguyen

**Reply by laggycomputer** [reply]
I will continue to support a timestamp in the interest of allowing AA and other consumers to recover from downtime. Imagine a course quickly went from full to waitlisted to full. If no request was made during that brief middle period, AA would never have realized that the course was waitlisted at all. Specifying a timestamp will allow you to discover this.

**Comment by laggycomputer** [comment]
@IsaacNguyen @MinhxNguyen7 I am working on a draft of the endpoint and there is a serious performance concern given your proposed use case of thousands of section codes. I currently see 4-6ms turnaround time per section code.

Are you willing to revise the contract of the endpoint to not guarantee any course or section data except the section code? If I understand your use case correctly you have no need to hear from us about that information in this endpoint.

**Reply by IsaacNguyen** [reply]
In all honesty, the only information we **really** need is the timestamp, department code, course title, course number, section code, restriction codes, and obviously the status. The number of people enrolled and that other information isn't exactly necessary for our use case. Does that sound about right @MinhxNguyen7?

**Reply by laggycomputer** [reply]
I did mean to say that I am interested in removing the department name and course. The restriction codes are considered part of enrollment status and will be part of the `from` and `to` objects.

**Reply by IsaacNguyen** [reply]
Though this doesn't exactly involve AAPI, I'm going to bring this up in this thread just for documentation sake. @MinhxNguyen7 do you think we could just store the department name and course on our end in the database at the initial creation of the subscription? For instance, when a user clicks on the notification bell for say ICS 53, the name, timeslot, etc. are all stored into the subscription at its creation. In this way, it would reduce the load of AAPI having to repeatedly scrape this information. Does this make sense?

**Reply by andrew-wang0** [reply]
I'll jump in here, I'm guessing not having to compute joins for each and every section to it's associated course would reduce the workload of our query by quite a lot, but having AntAlmanac store a copy would in essence just sidestep the work from our end to your end.

@laggycomputer if the query with joins comes out to be too computationally heavy, we can lock the endpoint behind a token and  return the course data with the sections.

**Reply by laggycomputer** [reply]
Thanks for the input. Currently trying about 4-5 different ways to go about doing this endpoint, I'll let you know what comes out.

**Reply by MinhxNguyen7** [reply]
@andrew-wang0 @laggycomputer @IsaacNguyen 

We only need the section code but not the department/course name because people will subscribe on a per-section basis. That was my imagination of the feature; I don't know if that makes it easier on AAPI. Does that sound about right?

As for the statuses changing by time, it would be quite nice to have that, but I would assume it'd necessitate expensive joins. We don't need to have it, and the contract can just be since the last scrape. If we have more than a few minutes of downtime, then the system might just miss some things. It is what it is. AAPI can, just as well, go down too.

**Reply by IsaacNguyen** [reply]
@MinhxNguyen7  

I only mentioned the part about the department/course name because when we send the actual email, we'll need those pieces of information to notify the user what course we are actually notifying them for. However, I think it's a good idea to just store that information in with the subscription itself upon its initial creation. An alternative would be calling AAPI one last time to gather that information about the section code before we send it out to the user. Do you have an idea on which would be more efficient?

**Reply by MinhxNguyen7** [reply]
I wouldn't want to store information that the API already has. If the endpoint itself doesn't have the information, we can just query for it. 

Especially if we are doing that (but as a general point also), you should think about parallelism/asynchronous tasks instead of sequentially waiting for information and sending emails.

@andrew-wang0 what do you think about that?

**Reply by andrew-wang0** [reply]
Agreed here, we'll want to avoid duplicating data between AA and AAPI. Though, relying on a secondary query to retrieve the necessary course information isn't the perfect solution either.

@laggycomputer worked extensively with the db logic for this, @laggycomputer can you confirm that we tested joining the course data for each section in the enrollment changes endpoint and found that to be infeasible? If that's the case we can continue with the plan for AA to query for course info.

**Comment by laggycomputer** [comment]
As I told you all on Discord, I have found a solution which provides only section codes. Each section object only has `sectionCode`, `from`, and `to` fields.

**Comment by laggycomputer** [comment]
PR is in review now: https://github.com/icssc/anteater-api/pull/125

**Reply by MinhxNguyen7** [reply]
@IsaacNguyen I believe that you can try out the endpoint on staging now, so you can see if it works for you.
